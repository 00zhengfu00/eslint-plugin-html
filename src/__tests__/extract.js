/*eslint-env es6*/
/*eslint no-sparse-arrays: 0*/

"use strict"
const extract = require("../extract")

function dedent(str) {
  if (str[0] === "\n") str = str.slice(1)

  const indent = str.match(/([\t ]*)\S/)
  if (indent) {
    str = str.replace(new RegExp(`^${indent[1]}`, "mg"), "")
  }

  // Remove the last empty line
  str = str.replace(/(\r\n|\r|\n)[\t ]*$/, "")

  return str
}

function placeholder(n) {
  return `/* HTML BLOCK ${n} */`
}

function test(params) {
  const infos = extract(
    dedent(params.input),
    params.indent,
    params.xmlMode,
    params.isJavaScriptMIMEType
  )
  expect(infos.code.toString()).toBe(dedent(params.output))
  expect(infos.badIndentationLines).toEqual(params.badIndentationLines || [])
}

it("extract simple javascript", () => {
  test({
    input: `
      some html
      <script>var foo = 1;</script>
      other
    `,
    output: `
      ${placeholder(0)}var foo = 1;${placeholder(1)}
    `,
  })
})

it("extract indented javascript", () => {
  test({
    input: `
      some html
      <script>
        var foo = 1;
      </script>
      other
    `,
    output: `
      ${placeholder(0)}
      var foo = 1;
      ${placeholder(1)}
    `,
  })
})

it("extract javascript with first line next to the script tag", () => {
  test({
    input: `
      some html
      <script>var foo = 1;
        var baz = 1;
      </script>
      other
    `,
    output: `
      ${placeholder(0)}var foo = 1;
      var baz = 1;
      ${placeholder(1)}
    `,
  })
})

it("extract javascript with last line next to the script tag", () => {
  test({
    input: `
      some html
      <script>
        var foo = 1;
        var baz = 1;</script>
      other
    `,
    output: `
      ${placeholder(0)}
      var foo = 1;
      var baz = 1;${placeholder(1)}
    `,
  })
})

it("extract multiple script tags", () => {
  test({
    input: `
      some html
      <script>
        var foo = 1;
      </script>
      other
      <script>
        var bar = 1;
      </script>
    `,
    output: `
      ${placeholder(0)}
      var foo = 1;
      ${placeholder(1)}
      var bar = 1;
      ${placeholder(2)}
    `,
  })
})

it("trim last line spaces", () => {
  test({
    input: `
      some html
        <script>
          var foo = 1;
        </script>
      other
    `,
    output: `
      ${placeholder(0)}
      var foo = 1;
      ${placeholder(1)}
    `,
  })
})

it("extract script containing 'lower than' characters correctly (#1)", () => {
  test({
    input: `
      <script>
        if (a < b) { doit(); }
      </script>
    `,
    output: `
      ${placeholder(0)}
      if (a < b) { doit(); }
      ${placeholder(1)}
    `,
  })
})


it("extract empty script tag (#7)", () => {
  test({
    input: `
      <script></script>
    `,
    output: `${placeholder(0)}${placeholder(1)}`,
  })
})

const prefixes = ["text/",
                "text/x-",
                "application/",
                "application/x-"]

const types = ["javascript", "babel"]

for (const prefix of prefixes) {
  for (const type of types) {
    const tag = `${prefix}${type}`

    it(`extracts a script tag with type=${tag}`, () => {
      test({
        input: `
          some html
          <script type="${tag}">var foo = 1;</script>
          other
        `,
        output: `
          ${placeholder(0)}var foo = 1;${placeholder(1)}
        `,
      })
    })
  }
}

it("collects bad indentations", () => {
  test({
    input: `
      <script>
        a;
      a;
       a;
      </script>
    `,
    output: `
      ${placeholder(0)}
      a;
      a;
       a;
      ${placeholder(1)}
    `,
    badIndentationLines: [ 3, 4 ],
  })
})

describe("indent option", () => {
  it("absolute indent with spaces", () => {
    test({
      input: `
        <head>
          <script>
            a;
          a;
        a;
          </script>
        </head>
      `,
      indent: {
        spaces: "  ",
      },
      output: `
        ${placeholder(0)}
            a;
        a;
        a;
        ${placeholder(1)}
      `,
      badIndentationLines: [ 3, 5 ],
    })
  })

  it("relative indent with spaces", () => {
    test({
      input: `
        <head>
          <script>
            a;
          a;
        a;
          </script>
        </head>
      `,
      indent: {
        spaces: "  ",
        relative: true,
      },
      output: `
        ${placeholder(0)}
        a;
          a;
        a;
        ${placeholder(1)}
      `,
      badIndentationLines: [ 4, 5 ],
    })
  })

  it("absolute indent with tabs", () => {
    test({
      input: `
        <head>
        \t<script>
        \t\ta;
        \ta;
        a;
        \t</script>
        </head>
      `,
      indent: {
        spaces: "\t",
      },
      output: `
        ${placeholder(0)}
        \t\ta;
        a;
        a;
        ${placeholder(1)}
      `,
      badIndentationLines: [ 3, 5 ],
    })
  })

  it("relative indent with tabs", () => {
    test({
      input: `
        <head>
        \t<script>
        \t\ta;
        \ta;
        a;
        \t</script>
        </head>
      `,
      indent: {
        spaces: "\t",
        relative: true,
      },
      output: `
        ${placeholder(0)}
        a;
        \ta;
        a;
        ${placeholder(1)}
      `,
      badIndentationLines: [ 4, 5 ],
    })
  })
})

it("works with crlf new lines", () => {
  test({
    input: "<p>\r\n</p>\r\n<script>\r\n  foo;\r\nbar;\r\n    baz;\r\n</script>\r\n",
    output: `${placeholder(0)}\r\nfoo;\r\nbar;\r\n  baz;\r\n${placeholder(1)}`,
    badIndentationLines: [ 5 ],
  })
})

it("works with CDATA", () => {
  test({
    input: `
    <script>
      a;
      <![CDATA[
      b;
      ]]>
      c;
    </script>`,
    xmlMode: true,
    output: `
      ${placeholder(0)}
      a;
      ${placeholder(1)}
      b;
      ${placeholder(2)}
      c;
      ${placeholder(3)}
    `,
  })
})

it("handles the isJavaScriptMIMEType option", () => {
  test({
    input: `
    <script>
      a
    </script>

    <script type="foo/bar">
      b
    </script>

    <script type="foo/baz">
      c
    </script>
    `,
    isJavaScriptMIMEType(type) { return type === "foo/bar" },
    output: `
    ${placeholder(0)}
    a
    ${placeholder(1)}
    b
    ${placeholder(2)}
    `,
  })
})
