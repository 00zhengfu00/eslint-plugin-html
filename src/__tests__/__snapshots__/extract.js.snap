// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collects bad indentations 1`] = `
"/* HTML */
a;
a;
 a;
/* HTML */
"
`;

exports[`extract empty script tag (#7) 1`] = `
"/* HTML *//* HTML */
"
`;

exports[`extract indented javascript 1`] = `
"/* HTML */
var foo = 1;
/* HTML */
"
`;

exports[`extract javascript with first line next to the script tag 1`] = `
"/* HTML */var foo = 1;
var baz = 1;
/* HTML */
"
`;

exports[`extract javascript with last line next to the script tag 1`] = `
"/* HTML */
var foo = 1;
var baz = 1;/* HTML */
"
`;

exports[`extract multiple script tags 1`] = `
"/* HTML */
var foo = 1;
/* HTML */
var bar = 1;
/* HTML */
"
`;

exports[`extract script containing 'lower than' characters correctly (#1) 1`] = `
"/* HTML */
if (a < b) { doit(); }
/* HTML */
"
`;

exports[`extract simple javascript 1`] = `
"/* HTML */var foo = 1;/* HTML */
"
`;

exports[`extracts a script tag with type=application/babel 1`] = `
"/* HTML */var foo = 1;/* HTML */
"
`;

exports[`extracts a script tag with type=application/javascript 1`] = `
"/* HTML */var foo = 1;/* HTML */
"
`;

exports[`extracts a script tag with type=application/x-babel 1`] = `
"/* HTML */var foo = 1;/* HTML */
"
`;

exports[`extracts a script tag with type=application/x-javascript 1`] = `
"/* HTML */var foo = 1;/* HTML */
"
`;

exports[`extracts a script tag with type=text/babel 1`] = `
"/* HTML */var foo = 1;/* HTML */
"
`;

exports[`extracts a script tag with type=text/javascript 1`] = `
"/* HTML */var foo = 1;/* HTML */
"
`;

exports[`extracts a script tag with type=text/x-babel 1`] = `
"/* HTML */var foo = 1;/* HTML */
"
`;

exports[`extracts a script tag with type=text/x-javascript 1`] = `
"/* HTML */var foo = 1;/* HTML */
"
`;

exports[`handles empty input 1`] = `""`;

exports[`handles self closing script tags in xhtml mode 1`] = `"/* HTML */"`;

exports[`handles the isJavaScriptMIMEType option 1`] = `
"/* HTML */
a
/* HTML */
b
/* HTML */
"
`;

exports[`indent option absolute indent with spaces 1`] = `
"/* HTML */
    a;
a;
a;
/* HTML */
"
`;

exports[`indent option absolute indent with tabs 1`] = `
"/* HTML */
		a;
a;
a;
/* HTML */
"
`;

exports[`indent option relative indent with spaces 1`] = `
"/* HTML */
a;
  a;
a;
/* HTML */
"
`;

exports[`indent option relative indent with tabs 1`] = `
"/* HTML */
a;
	a;
a;
/* HTML */
"
`;

exports[`keeps empty lines after the last html tags 1`] = `
"/* HTML */
a
/* HTML */


"
`;

exports[`trim last line spaces 1`] = `
"/* HTML */
var foo = 1;
/* HTML */
"
`;

exports[`works with CDATA 1`] = `
"/* HTML */
a;
/* HTML */
b;
/* HTML */
c;
/* HTML */"
`;

exports[`works with crlf new lines 1`] = `
"/* HTML */
foo;
bar;
  baz;
/* HTML */
"
`;
